File: js/src/model-manager.js
const fs = require('fs');
const path = require('path');
const os = require('os');
const https = require('https');

/**
 * The exact Hugging Face revision this release decodes.
 *
 * It used to be `main`, which is a moving ref: the artifact behind it has
 * already been replaced at least once (57.8 MB / 225 classes / height 64 ->
 * 26.4 MB / 316 classes / height 128), so two users running the same published
 * npm version could get two different networks. A commit is immutable; the
 * charset is fetched from the same one, so weights and vocabulary cannot come
 * from different versions of the repository.
 *
 * Bumping this is a deliberate act: change the revision, re-check the class
 * count and input height against `js/test/`, and update the bundled
 * `src/charset.txt` in the same commit.
 */
const MODEL_REVISION = 'a51be11';
const MODEL_REPO = 'janakhpon/monocr';

/**
 * Files fetched from the pinned revision. `bytes` is the size Hugging Face
 * reports for that revision; a file on disk of any other size is a partial or
 * corrupted download, not a cache hit.
 */
const REMOTE_FILES = [
    { remote: 'onnx/monocr.onnx', local: 'monocr.onnx', bytes: 26355440, label: 'model' },
    { remote: 'onnx/charset.txt', local: 'charset.txt', bytes: 674, label: 'charset' },
];

class ModelManager {
    constructor(revision = MODEL_REVISION) {
        this.revision = revision;

        // Cache directory in the user's home, keyed by revision.
        //
        // The old layout was a flat `~/.monocr/models/monocr.onnx` guarded by a
        // bare `fs.existsSync`, which cannot see that the file behind it came
        // from a different revision -- a stale artifact stays "cached" forever.
        // Putting the revision in the path makes a revision change a cache miss
        // by construction.
        this.rootDir = path.join(os.homedir(), '.monocr', 'models');
        this.cacheDir = path.join(this.rootDir, this.revision);

        this.baseUrl = `https://huggingface.co/${MODEL_REPO}/resolve/${this.revision}`;
        this.modelFileName = 'monocr.onnx';
        this.charsetFileName = 'charset.txt';
    }

    /**
     * Ensure cache directory exists
     */
    ensureCacheDir() {
        if (!fs.existsSync(this.cacheDir)) {
            fs.mkdirSync(this.cacheDir, { recursive: true });
        }
    }

    /**
     * Get local path for the model
     */
    getModelPath() {
        return path.join(this.cacheDir, this.modelFileName);
    }

    /**
     * Get local path for the charset that shipped with these weights
     */
    getCharsetPath() {
        return path.join(this.cacheDir, this.charsetFileName);
    }

    /**
     * True when a cached file is present *and* the right size. Size alone is not
     * integrity, but it does catch the failure this cache actually suffers:
     * an interrupted download left behind as a permanent, silently truncated
     * "cache hit".
     */
    hasFile(spec) {
        const target = path.join(this.cacheDir, spec.local);
        if (!fs.existsSync(target)) return false;
        return fs.statSync(target).size === spec.bytes;
    }

    /**
     * Check if the pinned revision is fully cached
     */
    hasModel() {
        return REMOTE_FILES.every((spec) => this.hasFile(spec));
    }

    /**
     * Download a file from HuggingFace.
     *
     * Writes to `<dest>.part` and renames on success, so an interrupted or
     * short download never occupies the cached path.
     */
    async downloadFile(url, destPath, expectedBytes, label) {
        const tmpPath = `${destPath}.part`;

        await new Promise((resolve, reject) => {
            const file = fs.createWriteStream(tmpPath);
            let settled = false;
            const fail = (err) => {
                if (settled) return;
                settled = true;
                file.destroy();
                fs.unlink(tmpPath, () => reject(err));
            };

            const request = (requestUrl, redirectsLeft) => {
                https.get(requestUrl, { headers: { 'User-Agent': 'monocr-npm' } }, (response) => {
                    if ([301, 302, 307, 308].includes(response.statusCode)) {
                        if (redirectsLeft <= 0) {
                            response.resume();
                            fail(new Error(`Too many redirects fetching ${url}`));
                            return;
                        }
                        let redirectUrl = response.headers.location;
                        if (!redirectUrl.startsWith('http')) {
                            const originalUrl = new URL(requestUrl);
                            redirectUrl = `${originalUrl.protocol}//${originalUrl.host}${redirectUrl}`;
                        }
                        response.resume();
                        request(redirectUrl, redirectsLeft - 1);
                    } else if (response.statusCode === 200) {
                        const totalSize = parseInt(response.headers['content-length'], 10);
                        let downloadedSize = 0;

                        response.on('data', (chunk) => {
                            downloadedSize += chunk.length;
                            if (totalSize && process.stdout.isTTY) {
                                const progress = ((downloadedSize / totalSize) * 100).toFixed(1);
                                process.stdout.write(`\r  Downloading ${label}: ${progress}% (${(downloadedSize / 1024 / 1024).toFixed(2)} MB)`);
                            }
                        });
                        response.on('error', fail);

                        response.pipe(file);

                        file.on('finish', () => {
                            file.close((err) => {
                                if (err) return fail(err);
                                if (process.stdout.isTTY && totalSize) process.stdout.write('\n');
                                if (settled) return;
                                settled = true;
                                resolve();
                            });
                        });
                    } else {
                        response.resume();
                        fail(new Error(`Failed to download ${url}: HTTP ${response.statusCode}`));
                    }
                }).on('error', fail);
            };

            file.on('error', fail);
            request(url, 5);
        });

        const actualBytes = fs.statSync(tmpPath).size;
        if (typeof expectedBytes === 'number' && actualBytes !== expectedBytes) {
            fs.unlinkSync(tmpPath);
            throw new Error(
                `Downloaded ${label} is ${actualBytes} bytes, expected ${expectedBytes} ` +
                `at revision ${this.revision}. Refusing to cache it.`
            );
        }

        fs.renameSync(tmpPath, destPath);
    }

    /**
     * Download the model and the charset that belongs to it
     */
    async downloadModel() {
        this.ensureCacheDir();
        this.warnAboutLegacyCache();

        console.log(`Downloading monocr model from HuggingFace (${MODEL_REPO} @ ${this.revision})...`);
        console.log(`Cache directory: ${this.cacheDir}`);

        let fetched = 0;
        for (const spec of REMOTE_FILES) {
            if (this.hasFile(spec)) continue;
            await this.downloadFile(
                `${this.baseUrl}/${spec.remote}`,
                path.join(this.cacheDir, spec.local),
                spec.bytes,
                spec.label
            );
            fetched += 1;
        }

        console.log(fetched === 0 ? '✓ Already cached.' : '✓ Model downloaded successfully!');
    }

    /**
     * Point at the flat pre-revision cache if it is still lying around. It is
     * dead weight now (~58 MB) and it belongs to a network this code cannot
     * decode, but it is the user's file, so say so rather than delete it.
     */
    warnAboutLegacyCache() {
        const legacy = path.join(this.rootDir, this.modelFileName);
        try {
            if (fs.existsSync(legacy) && fs.statSync(legacy).isFile()) {
                console.warn(
                    `Note: ${legacy} is from an older, unpinned download and is no longer used. ` +
                    'You can delete it.'
                );
            }
        } catch (err) {
            // Never let a cache-hygiene notice break a download.
        }
    }

    /**
     * Get the cached artifacts, downloading if needed.
     *
     * @returns {Promise<{modelPath: string, charsetPath: string, revision: string}>}
     */
    async ensureModel() {
        if (!this.hasModel()) {
            await this.downloadModel();
        }
        return {
            modelPath: this.getModelPath(),
            charsetPath: this.getCharsetPath(),
            revision: this.revision,
        };
    }
}

module.exports = ModelManager;
module.exports.MODEL_REVISION = MODEL_REVISION;
module.exports.MODEL_REPO = MODEL_REPO;
module.exports.REMOTE_FILES = REMOTE_FILES;

File: rust/src/model_manager.rs
//! Model Manager
//!
//! This module handles downloading and caching the ONNX model used for OCR.
//! Models are downloaded from HuggingFace and stored in the user's cache directory.

use indicatif::{ProgressBar, ProgressStyle};
use reqwest::blocking::Client;
use std::fs;
use std::io;
use std::path::{Path, PathBuf};

/// The Hugging Face repository holding the ONNX artifact.
pub const MODEL_REPO: &str = "janakhpon/monocr";

/// The pinned revision.
///
/// `main` is a moving ref and the artifact has already changed under it: the
/// model served at one point had a 64-pixel input and 225 output classes, the
/// one served now has 128 and 316. A cache that gates on "the file exists"
/// cannot tell those apart, so the revision is part of the cache path.
pub const MODEL_REVISION: &str = "d3d9d5e";

/// Filename of the ONNX model within the repository's `onnx/` directory.
pub const MODEL_FILENAME: &str = "monocr.onnx";

/// Filename of the charset that belongs to the same revision.
pub const CHARSET_FILENAME: &str = "charset.txt";

/// Manages downloading and caching of OCR models
///
/// This struct handles the lifecycle of the ONNX model file, including:
/// - Determining the cache location (`~/.monocr/models/<revision>/`)
/// - Downloading the model and its charset from HuggingFace if not present
/// - Providing the path to the model file for loading
pub struct ModelManager {
    /// Directory where this revision's files are cached
    cache_dir: PathBuf,
    /// Base URL for downloading, already pinned to a revision
    base_url: String,
    /// Filename of the model file
    model_filename: String,
}

impl Default for ModelManager {
    fn default() -> Self {
        Self::new()
    }
}

impl ModelManager {
    /// Create a new ModelManager with default settings
    ///
    /// # Default Values
    ///
    /// - Cache directory: `~/.monocr/models/<revision>/`
    /// - Base URL: `https://huggingface.co/janakhpon/monocr/resolve/<revision>`
    /// - Model filename: `monocr.onnx`
    ///
    /// # Panics
    ///
    /// Panics if the user's home directory cannot be determined
    pub fn new() -> Self {
        let home = dirs::home_dir().expect("Failed to get home directory");
        // Revision-scoped: re-pinning MODEL_REVISION is a cache miss rather
        // than a silent reuse of whatever was downloaded last time.
        let cache_dir = home.join(".monocr").join("models").join(MODEL_REVISION);

        Self {
            cache_dir,
            base_url: format!("https://huggingface.co/{MODEL_REPO}/resolve/{MODEL_REVISION}"),
            model_filename: MODEL_FILENAME.to_string(),
        }
    }

    /// The directory this manager downloads into.
    pub fn cache_dir(&self) -> &Path {
        &self.cache_dir
    }

    /// The pinned download URL for the ONNX model.
    pub fn model_url(&self) -> String {
        format!("{}/onnx/{}", self.base_url, self.model_filename)
    }

    /// The pinned download URL for the charset.
    ///
    /// Same revision as the weights — that is the only way to be sure the two
    /// agree.
    pub fn charset_url(&self) -> String {
        format!("{}/onnx/{}", self.base_url, CHARSET_FILENAME)
    }

    /// Get the path to the ONNX model file, downloading it if the cache for
    /// this revision is empty.
    pub fn get_model_path(&self) -> io::Result<PathBuf> {
        let model_path = self.cache_dir.join(&self.model_filename);

        if !model_path.exists() {
            println!(
                "Model {MODEL_REVISION} not found at {:?}. Downloading...",
                model_path
            );
            self.download(&self.model_url(), &model_path)?;
            println!("Download complete");
        }

        Ok(model_path)
    }

    /// Get the charset published alongside the pinned model.
    ///
    /// Preferred over the embedded copy because it comes from the same revision
    /// as the weights.
    pub fn get_charset(&self) -> io::Result<String> {
        let charset_path = self.cache_dir.join(CHARSET_FILENAME);

        if !charset_path.exists() {
            self.download(&self.charset_url(), &charset_path)?;
        }

        fs::read_to_string(&charset_path)
    }

    /// Download `url` to `dest` via a temporary file.
    ///
    /// The rename is the last step, so an interrupted transfer never leaves a
    /// truncated artifact behind for the existence check to accept.
    fn download(&self, url: &str, dest: &Path) -> io::Result<()> {
        if let Some(parent) = dest.parent() {
            fs::create_dir_all(parent)?;
        }

        let client = Client::new();
        let mut response = client
            .get(url)
            .send()
            .map_err(|e| io::Error::other(format!("Failed to download from {}: {}", url, e)))?;

        if !response.status().is_success() {
            return Err(io::Error::other(format!(
                "Failed to download {}: {}",
                url,
                response.status()
            )));
        }

        let total_size = response.content_length().unwrap_or(0);
        let pb = ProgressBar::new(total_size);
        pb.set_style(ProgressStyle::default_bar()
            .template("{spinner:.green} [{elapsed_precise}] [{wide_bar:.cyan/blue}] {bytes}/{total_bytes} ({eta})")
            .unwrap()
            .progress_chars(">-"));

        let tmp_path = dest.with_extension("part");
        let mut file = fs::File::create(&tmp_path)?;
        let copied = io::copy(&mut response, &mut file);
        drop(file);

        if let Err(e) = copied {
            let _ = fs::remove_file(&tmp_path);
            return Err(e);
        }

        fs::rename(&tmp_path, dest)?;
        pb.finish_and_clear();
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// `main` is a moving ref and the artifact has already changed under it.
    #[test]
    fn download_urls_are_pinned() {
        let m = ModelManager::new();
        for url in [m.model_url(), m.charset_url()] {
            assert!(
                !url.contains("/resolve/main/"),
                "still tracking the moving ref `main`: {url}"
            );
            assert!(
                url.contains(&format!("/resolve/{MODEL_REVISION}/")),
                "not pinned to {MODEL_REVISION}: {url}"
            );
        }
    }

    /// The charset has to come from the same revision as the weights.
    #[test]
    fn charset_is_fetched_from_the_model_revision() {
        let m = ModelManager::new();
        let model_dir = m.model_url().trim_end_matches(MODEL_FILENAME).to_string();
        let charset_dir = m
            .charset_url()
            .trim_end_matches(CHARSET_FILENAME)
            .to_string();
        assert_eq!(model_dir, charset_dir);
    }

    /// The cache used to gate on file existence alone, so an artifact from an
    /// older revision was reused forever.
    #[test]
    fn cache_dir_is_scoped_by_revision() {
        let m = ModelManager::new();
        assert_eq!(
            m.cache_dir().file_name().and_then(|s| s.to_str()),
            Some(MODEL_REVISION),
            "cache directory {:?} is not revision-scoped",
            m.cache_dir()
        );
    }
}
